content/post/2019-11-26.Rmd

```diff
@@ -0,0 +1,640 @@
+should return the product of all positive integers less than or equal to _n_
+
+Step 1 (RED).  Write a test:
+```{r error=TRUE}
+# tests/testthat/test_myfactorial.R
+test_that("Valid arguments", {
+    expect_equal(my_factorial(0), 1)
+    expect_equal(my_factorial(3), 6)
+})
+```
+
+
+Step 2 (GREEN).  Write the minimum amount of code that makes the tests pass:
+```{r error=TRUE}
+# R/myfactorial.R
+my_factorial <- function(n) {
+    ifelse(n == 0, 1, 6)
+    # Hey! This definitely looks like cheating! What are you up to?
+}
+
+# tests/testthat/test_myfactorial.R
+test_that("Valid arguments", {
+    expect_equal(my_factorial(0), 1)
+    expect_equal(my_factorial(3), 6)
+})
+```
+
+Test passing now.
+
+
+Step 3 (RED).  Write yet another test:
+```{r error=TRUE}
+# tests/testthat/test_myfactorial.R
+test_that("Valid arguments", {
+    expect_equal(my_factorial(0), 1)
+    expect_equal(my_factorial(3), 6)
+    expect_equal(my_factorial(10), 3628800)
+})
+```
+
+
+Step 4 (GREEN+REFACTOR).  We don't have a choice but to refactor. The tests already in place will keep our new code true!
+
+```{r error=TRUE}
+# R/myfactorial.R
+my_factorial <- function(n) {
+    if(n == 0) return(1)
+
+    factorial <- 1
+    for(i in 1:n) {
+        factorial <- i * factorial
+    }
+    return(factorial)
+}
+
+# tests/testthat/test_myfactorial.R
+test_that("Valid arguments", {
+    expect_equal(my_factorial(0), 1)
+    expect_equal(my_factorial(3), 6)
+    expect_equal(my_factorial(10), 3628800)
+})
+```
+
+Test passing now.
+
+
+The final two requirements are left as an exercise to the reader (go on, don't be shy!).
+
+...
+
+The final result might look something like this:
+
+```{r error=TRUE}
+# R/myfactorial.R
+my_factorial <- function(n) {
+    stopifnot(is.numeric(n))
+    if(n < 0 | n > 14) return(NULL)
+    if(n == 0) return(1)
+
+    factorial <- 1
+    for(i in 1:n) {
+        factorial <- i * factorial
+    }
+    return(factorial)
+}
+
+# tests/testthat/test_myfactorial.R
+
+test_that("Valid arguments", {
+    expect_equal(my_factorial(0), 1)
+    expect_equal(my_factorial(3), 6)
+    expect_equal(my_factorial(10), 3628800)
+
+    rnd_num <- sample.int(14, 1)
+    expect_is(my_factorial(rnd_num), "numeric")
+})
+
+test_that("Null for negatives", {
+    neg_num <- -sample.int(100, 1)
+    expect_equal(my_factorial(neg_num), NULL)
+})
+
+test_that("Null for numbers bigger than 14", {
+    num_bigger_than_14 <- 14 + sample.int(100, 1)
+    expect_equal(my_factorial(num_bigger_than_14), NULL)
+})
+```
+
+## Beyond Unit Testing
+
+### Integration
+CAVEAT EMPTOR!  Your code can pass all the unit tests, with a 100% coverage, and still have problems!
+
+<center>
+![](/post/2019-11-26_files/nointegrationitests.png)
+</center>
+
+These are called *integration* problems, which are tackled using _integration tests_.
+This form of testing comprises tests that, er, test more than one unit (read: function) at a time.
```

Comment:
No tengo muy claro que significa el 'er'. Supongo que será una abreviatura de "ergo", pero no estoy seguro. 

De todas formas, yo creo que la frase se entiende mejor sin él: ...comprises tests that test more... 

si no quieres repetir la plabra _test_ puedes usar analyze, check or verify.

Reply:
He sustituído el 'er' por 'ahem'. Intento aprovechar la redundancia para emplear un tono más conversacional.